Update sleep manager tests

fkjagodzinski opened the pull request:
### Description

Extended `tests-mbed_hal-sleep_manager` with:
* `deep sleep lock/unlock`,
* `deep sleep locked USHRT_MAX times`,
* `deep sleep locked more than USHRT_MAX times`,
* `deep sleep unbalanced unlock`,
* `sleep_auto calls sleep/deep sleep based on lock`,
*  `deep sleep lock/unlock test_check`.

Updated `tests-mbed_hal-sleep_manager` and `tests-mbed_hal-sleep_manager_racecondition` test case docs and moved them to test header files.

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [x] Feature
    [ ] Breaking change

CC @c1728p9 @mprse @jamesbeyond @0xc0170

fkjagodzinski committed:
Tests: SleepManager: extend test suite

New test cases:
* "deep sleep lock/unlock"
* "deep sleep unbalanced unlock"
* "deep sleep locked USHRT_MAX times"
* "deep sleep locked more than USHRT_MAX times"
* "sleep_auto calls sleep/deep sleep based on lock"

fkjagodzinski committed:
Tests: SleepManager: update concurrent test docs

fkjagodzinski committed:
HAL: SleepManager: Update header file

Move sleep manager API into its own doxygen group.
Add defined behavior section and links to tests.

fkjagodzinski commented on TESTS/mbed_hal/sleep_manager/main.cpp:
```diff
@@ -16,45 +16,169 @@
+{
+    GREENTEA_SETUP(10, "default_auto");
+    // Suspend the RTOS kernel scheduler to prevent any OS interference.
+    osKernelSuspend();
```
NOTE:
As @mprse pointed out, when https://github.com/ARMmbed/mbed-os/pull/7524 is merged, this wait should be added here.
```cpp
#if DEVICE_LPTICKER && (LPTICKER_DELAY_TICKS > 0)
    // After the OS is disabled wait until the microsecond timer
    // is no longer in use by the lp ticker wrapper code.
    // This prevents the low power ticker wrapper code from
    // getting stuck in a state where it is waiting for
    // the microsecond Timeout to run.
    while (lp_ticker_get_timeout_pending());
#endif
```

fkjagodzinski commented on platform/mbed_power_mgmt.h:
```diff
@@ -96,23 +96,32 @@ void sleep_tracker_unlock(const char *const filename, int line);
+ * Verified by ::test_lock_eq_ushrt_max and ::test_lock_gt_ushrt_max
+ *
  * This function is IRQ and thread safe
+ * Verified by ::sleep_manager_multithread_test and ::sleep_manager_irq_test
```
@0xc0170 Is this way of referring tests acceptable? Or should I rather insert a separate defined/undefined behaviour section, e.g. as seen in the ticker API? https://github.com/ARMmbed/mbed-os/blob/5d5ca62a5e6ee836b86503d58cae6eec8444458f/hal/us_ticker_api.h#L60-L87

0xc0170 replied:
I assume as its in ticker (defined behavior + how it is verified)

fkjagodzinski replied:
I created a new doxygen section for defined behavior.

0xc0170 commented on hal/mbed_sleep_manager.c:
```diff
@@ -162,6 +162,7 @@ void sleep_manager_lock_deep_sleep_internal(void)
     if (deep_sleep_lock == USHRT_MAX) {
         core_util_critical_section_exit();
         MBED_ERROR1(MBED_MAKE_ERROR(MBED_MODULE_HAL, MBED_ERROR_CODE_OVERFLOW), "DeepSleepLock overflow (> USHRT_MAX)", deep_sleep_lock);
+        return; // Handle overridden mbed_error(), i.e. when running sleep_manager tests.
```
This return was added so while running tests, it would immediately return instead of continue (does not trap)?

fkjagodzinski replied:
Yes, since I overridden `mbed_error()` for test purposes and now it does not halt, I had to add this return to prevent execution of following lines -- changing the value of `deep_sleep_lock` var.

0xc0170 commented on TESTS/mbed_hal/sleep_manager/main.cpp:
```diff
@@ -16,45 +16,169 @@
+    us_diff2 = ticker_read_us(us_ticker) - us_ts;
+    lp_diff2 = ticker_read_us(lp_ticker) - lp_ts;
+
+    TEST_ASSERT_MESSAGE(us_diff2 < lp_diff2 / 4, "Deep sleep mode not used when unlocked");
```
where does `4` come from? what are these 2 lines testing ?

fkjagodzinski replied:
Now I can see, I didn't provide much explanation here; will add more verbose comments.
This test relies on the fact that us tickers are powered off during deep sleep, and lp tickers stay on for both sleep and deepsleep. The type of sleep that was actually used can be checked by comparing time measured by us and lp tickers.

`4` is an arbitrary value. Test checks if us ticker incremented at most 25% of lp ticker time.

0xc0170 replied:
Please update with comments

fkjagodzinski replied:
Comments added.

0xc0170 commented:
Please review travis docs failure

fkjagodzinski commented:
Doxygen generates fine locally, so travis should also be happy with updated code.

cmonr commented:
@0xc0170 @bulislaw @ARMmbed/mbed-os-test Mind re-reviewing?

jamesbeyond reviewed (approved):
LGTM

fkjagodzinski commented:
@0xc0170 @bulislaw @c1728p9 could you take a look at this PR?

fkjagodzinski commented:
I noticed there was a conflict due to recent astyle update on TESTS dir, so I rebased changes onto current master and update the files using astyle locally.

Although there are no conflicts at the moment, I think this PR could wait for #7524. Both update sleep_manager test files but #7524 looks to be further in the pipeline.

0xc0170 commented:
> Although there are no conflicts at the moment, I think this PR could wait for #7524. Both update sleep_manager test files but #7524 looks to be further in the pipeline.

Already integrated.

Please rebase to resolve the conflict

fkjagodzinski committed:
Tests: SleepManager: Add test for new check() fun

sleep_manager_can_deep_sleep_test_check() is intended to be used in test
code.

fkjagodzinski commented:
* rebased to master,
* added `deep sleep lock/unlock test_check` tests for new `sleep_manager_can_deep_sleep_test_check()` function from #7524.

fkjagodzinski committed:
Tests: SleepManager: Fix test suite setup

Use ticker init functions from upper HAL layer to allow proper handling
of ticker overflows.

fkjagodzinski commented:
I had some problems with us ticker on NRF51_DK not returning the correct time during ordinary sleep(). Investigation revealed that I used a function from wrong HAL layer to init the us ticker, and as a consequence this ticker was not protected against overflow when sleeping. This issue was discovered on NRF51_DK which uses a 16 bit counter for us ticker.

The most recent commit fixes this issue. I think this PR is ready for morph tests.

0xc0170 commented:
/morph build

0xc0170 commented:
/morph build

studavekar commented:
> Test : FAILURE
> Build number : 2778
> Test logs :http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/7582/2778

Looks like nucleo device went in bad state running the tests.

Failure for NUCLEO_F746ZG
```
[1536157327.04][CONN][RXD] >>> Running case #5: 'sleep_auto calls sleep/deep sleep based on lock'...
[1536157327.14][CONN][INF] found KV pair in stream: {{__testcase_start;sleep_auto calls sleep/deep sleep based on lock}}, queued...
[1536157327.24][CONN][RXD] :155::FAIL: Expected TRUE Was FALSE
[1536157327.34][CONN][INF] found KV pair in stream: {{__testcase_finish;sleep_auto calls sleep/deep sleep based on lock;0;1}}, queued...
[1536157327.44][CONN][RXD] >>> 'sleep_auto calls sleep/deep sleep based on lock': 0 passed, 1 failed with reason 'Assertion Failed'
```

cmonr commented:
Holding off on restarting Test CI to prioritize other PRs needed for releases happening this week.

fkjagodzinski commented:
Regarding `sleep_auto calls sleep/deep sleep based on lock` test case that failed on `NUCLEO_F401RE` and `NUCLEO_F746ZG` -- I can see that HAL sleep calls are made correctly. `hal_deepsleep()` and `hal_sleep()` are called as expected. However, there seems to be an issue with value retuned by `ticker_read_us(us_ticker)` after deep sleep. I think I'll update this test case to use ticker counters directly, i.e. read values via `us_ticker_read()`.

fkjagodzinski commented:
I created an issue #8117 regarding the behavior of some STM targets.

fkjagodzinski commented:
Waiting for #8136

fkjagodzinski commented:
I can see #8136 has just been merged, which means this PR is ready to run tests again.

cmonr commented:
/morph build

cmonr commented:
@fkjagodzinski Please take a look at the test failures

fkjagodzinski committed:
Tests: SleepManager: Update test_check() test

Increase delta to 500 us. This value is still short enough to detect
incorrect behavior of the sleep_manager_can_deep_sleep_test_check() fun,
but allows the targets with low LP timer accuracy to pass, i.e.
NUCLEO_F429ZI.

fkjagodzinski commented:
Regarding latest CI failures for this PR:
* `deep sleep lock/unlock test_check` that failed on `NUCLEO_F429ZI` is now fixed with 
77f9f25,
* `sleep_auto calls sleep/deep sleep based on lock` that failed on `NUCLEO_F746ZG` is a result of issue #8278.

The issue mentioned above should be fixed by #8279.
I also rebased the code onto current master to keep the PR up to date.

cmonr commented:
/morph build

fkjagodzinski commented:
Waiting for PR #8279 which will fix the last issue.

fkjagodzinski committed:
Tests: Sleep: Extract test utility functions

Move a few of utility functions so other tests can use them.

fkjagodzinski committed:
Tests: SleepManager: Fix sleep_auto test

Use us & lp tickers directly, without the common ticker layer.

fkjagodzinski commented:
Update:
* rebased to master,
* updated `sleep_auto calls sleep/deep sleep based on lock` test case to use lp and us tickers directly, with upper ticker layer and `LowPowerTickerWrapper` suspended -- no need to wait for the outcome of #8278 

PR is ready for morph tests again.

0xc0170 commented:
/morph build
